File: apps/web/src/services/pricing.ts
/**
 * 会员定价服务 - 从后端获取定价信息
 */

import { api } from './api'

// 计费周期类型
export type BillingCycle = 'monthly' | 'quarterly' | 'yearly'

// 会员等级类型
export type MembershipTier = 'basic' | 'vip' | 'svip'

// 价格详情
export interface PriceDetail {
  original_price: number      // 原价
  discounted_price: number    // 折扣价
  discount_rate: number       // 折扣率 (0.0 - 1.0)
  discount_label: string      // 折扣标签，如 "8折"
  save_amount: number         // 节省金额
}

// 会员方案
export interface MembershipPlan {
  tier: MembershipTier
  name: string
  description: string
  icon: string
  color: string
  badge?: string              // 标签，如 "最受欢迎"
  features: string[]          // 功能列表
  highlights: string[]        // 功能亮点
  prices: Record<BillingCycle, PriceDetail>
  is_popular?: boolean
}

// 定价配置响应
export interface PricingConfigResponse {
  plans: MembershipPlan[]
  cycles: {
    cycle: BillingCycle
    label: string
    unit: string
    months: number
  }[]
  global_discount?: {
    label: string
    rate: number
    valid_until: string
  }
}

// 创建订单请求
export interface CreateOrderRequest {
  tier: MembershipTier
  cycle: BillingCycle
  coupon_code?: string
}

// 创建订单响应
export interface CreateOrderResponse {
  order_id: string
  tier: MembershipTier
  cycle: BillingCycle
  original_amount: number
  discount_amount: number
  final_amount: number
  pay_url: string             // 支付链接（支付宝/微信）
  expires_at: string          // 订单过期时间
}

// 用户当前会员信息
export interface UserMembership {
  tier: MembershipTier
  status: 'active' | 'expired' | 'cancelled'
  start_date: string
  end_date: string
  auto_renew: boolean
  next_billing_date?: string
}

export const pricingApi = {
  /**
   * 获取定价配置
   */
  getPricingConfig: async (): Promise<PricingConfigResponse> => {
    const response = await api.get('/pricing/config')
    return response.data
  },

  /**
   * 获取当前用户的会员信息
   */
  getUserMembership: async (): Promise<UserMembership | null> => {
    try {
      const response = await api.get('/pricing/membership')
      return response.data
    } catch (error) {
      // 未登录或没有会员信息返回 null
      return null
    }
  },

  /**
   * 创建订阅订单
   * TODO: 支付对接完成后
   * 1. 根据返回的 pay_url 或 qr_code 调起支付
   * 2. 轮询支付状态或等待回调
   * 3. 支付成功后跳转会员中心
   */
  createOrder: async (data: CreateOrderRequest): Promise<CreateOrderResponse> => {
    const response = await api.post('/pricing/order', data)
    return response.data
  },

  /**
   * 验证优惠券
   * TODO: 优惠券系统对接
   * 应用场景:
   * 1. 用户在定价页面输入优惠券码
   * 2. 实时验证优惠券有效性
   * 3. 显示折扣后价格
   * 4. 创建订单时携带 coupon_code
   */
  validateCoupon: async (coupon_code: string, tier: MembershipTier, cycle: BillingCycle): Promise<{
    valid: boolean
    discount_amount: number
    message?: string
  }> => {
    const response = await api.post('/pricing/coupon/validate', {
      coupon_code,
      tier,
      cycle,
    })
    return response.data
  },

  /**
   * 获取支付状态
   * TODO: 支付对接完成后
   * 使用场景:
   * 1. 订单详情页轮询支付状态
   * 2. 支付完成后自动刷新页面
   * 3. 超时未支付提醒
   */
  getPaymentStatus: async (order_id: string): Promise<{
    status: 'pending' | 'paid' | 'failed' | 'cancelled'
    paid_at?: string
  }> => {
    const response = await api.get(`/pricing/order/${order_id}/status`)
    return response.data
  },

  /**
   * 取消订阅（自动续费）
   * TODO: 自动续费对接完成后
   * 使用场景:
   * 1. 会员中心页面关闭自动续费
   * 2. 调用支付宝/微信解约接口
   * 3. 更新本地 auto_renew 状态
   * 注意: 取消自动续费不影响当前会员有效期
   */
  cancelSubscription: async (): Promise<void> => {
    await api.post('/pricing/subscription/cancel')
  },
}

export default pricingApi

File: apps/api/internal/handler/pricing.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

// PricingHandler 定价处理器
type PricingHandler struct{}

// NewPricingHandler 创建定价处理器
func NewPricingHandler() *PricingHandler {
	return &PricingHandler{}
}

// PriceDetail 价格详情
type PriceDetail struct {
	OriginalPrice   float64 `json:"original_price"`
	DiscountedPrice float64 `json:"discounted_price"`
	DiscountRate    float64 `json:"discount_rate"`
	DiscountLabel   string  `json:"discount_label"`
	SaveAmount      float64 `json:"save_amount"`
}

// MembershipPlan 会员方案
type MembershipPlan struct {
	Tier        string                     `json:"tier"`
	Name        string                     `json:"name"`
	Description string                     `json:"description"`
	Icon        string                     `json:"icon"`
	Color       string                     `json:"color"`
	Badge       string                     `json:"badge,omitempty"`
	Features    []string                   `json:"features"`
	Highlights  []string                   `json:"highlights"`
	Prices      map[string]PriceDetail     `json:"prices"`
	IsPopular   bool                       `json:"is_popular"`
}

// PricingConfigResponse 定价配置响应
type PricingConfigResponse struct {
	Plans           []MembershipPlan `json:"plans"`
	Cycles          []CycleInfo      `json:"cycles"`
	GlobalDiscount  *GlobalDiscount  `json:"global_discount,omitempty"`
}

// CycleInfo 计费周期信息
type CycleInfo struct {
	Cycle   string `json:"cycle"`
	Label   string `json:"label"`
	Unit    string `json:"unit"`
	Months  int    `json:"months"`
}

// GetPricingConfig 获取定价配置
func (h *PricingHandler) GetPricingConfig(c *gin.Context) {
	config := PricingConfigResponse{
		Plans: []MembershipPlan{
			{
				Tier:        "basic",
				Name:        "基础版",
				Description: "适合个人投资者入门使用",
				Icon:        "star",
				Color:       "blue",
				Features: []string{
					"每日涨停预测信号",
					"基础股票筛选",
					" limited 回测功能",
					"社区讨论访问",
				},
				Highlights: []string{"免费使用"},
				Prices: map[string]PriceDetail{
					"monthly":    {OriginalPrice: 0, DiscountedPrice: 0, DiscountRate: 1, DiscountLabel: "", SaveAmount: 0},
					"quarterly":  {OriginalPrice: 0, DiscountedPrice: 0, DiscountRate: 1, DiscountLabel: "", SaveAmount: 0},
					"yearly":     {OriginalPrice: 0, DiscountedPrice: 0, DiscountRate: 1, DiscountLabel: "", SaveAmount: 0},
				},
				IsPopular: false,
			},
			{
				Tier:        "vip",
				Name:        "VIP会员",
				Description: "解锁全部高级分析功能",
				Icon:        "crown",
				Color:       "gold",
				Badge:       "最受欢迎",
				Features: []string{
					"全部基础版功能",
					"实时涨停信号推送",
					"AI 智能选股策略",
					"无限回测分析",
					"龙虎榜资金流向",
					"游资动向追踪",
					"专属客服支持",
				},
				Highlights: []string{"实时推送", "AI选股"},
				Prices: map[string]PriceDetail{
					"monthly":    {OriginalPrice: 99, DiscountedPrice: 99, DiscountRate: 1, DiscountLabel: "", SaveAmount: 0},
					"quarterly":  {OriginalPrice: 297, DiscountedPrice: 249, DiscountRate: 0.84, DiscountLabel: "84折", SaveAmount: 48},
					"yearly":     {OriginalPrice: 1188, DiscountedPrice: 799, DiscountRate: 0.67, DiscountLabel: "67折", SaveAmount: 389},
				},
				IsPopular: true,
			},
			{
				Tier:        "svip",
				Name:        "SVIP会员",
				Description: "机构级专业分析工具",
				Icon:        "thunderbolt",
				Color:       "purple",
				Features: []string{
					"全部VIP功能",
					"机构专用策略模型",
					"大宗交易监控",
					"主力资金流向分析",
					"产业链关联分析",
					"1对1投资顾问",
					"API接口访问",
					"优先体验新功能",
				},
				Highlights: []string{"机构策略", "1对1顾问"},
				Prices: map[string]PriceDetail{
					"monthly":    {OriginalPrice: 299, DiscountedPrice: 299, DiscountRate: 1, DiscountLabel: "", SaveAmount: 0},
					"quarterly":  {OriginalPrice: 897, DiscountedPrice: 749, DiscountRate: 0.83, DiscountLabel: "83折", SaveAmount: 148},
					"yearly":     {OriginalPrice: 3588, DiscountedPrice: 2399, DiscountRate: 0.67, DiscountLabel: "67折", SaveAmount: 1189},
				},
				IsPopular: false,
			},
		},
		Cycles: []CycleInfo{
			{Cycle: "monthly", Label: "月付", Unit: "/月", Months: 1},
			{Cycle: "quarterly", Label: "季付", Unit: "/季", Months: 3},
			{Cycle: "yearly", Label: "年付", Unit: "/年", Months: 12},
		},
	}

	c.JSON(http.StatusOK, config)
}
